diff --git a/apps/app_voicemail.c b/apps/app_voicemail.c
index 3c5fa8460d..16b52f343c 100644
--- a/apps/app_voicemail.c
+++ b/apps/app_voicemail.c
@@ -7210,7 +7210,7 @@ static int leave_voicemail(struct ast_channel *chan, char *ext, struct leave_vm_
 #endif
 		} else {
 			ast_trace(-1, "%s doesn't exist, doing what we can\n", prefile);
-			res = invent_message(chan, vmu->context, ext, ast_test_flag(options, OPT_BUSY_GREETING), ecodes);
+			res = 0;
 		}
 		SCOPE_CALL(-1, DISPOSE, prefile, -1);
 		if (res < 0) {
